File: CPP/Program/Vip.cpp
#pragma once
#include <iostream>
#include <string>
#include <iomanip>
#include "Customer.cpp"
#include "Membership.cpp"

// This class represents a VIP Customer, inheriting from Customer and Membership
class VIPCustomer : public Customer, public Membership {
    // Attributes specific to VIP Customer
    private:
        int totalSpent;
    
    public:
        // Default constructor
        VIPCustomer() : Customer(), Membership() {}

        // Parameterized constructor
        VIPCustomer(string name, int age, string gender, string customerId, string handphoneNumber, string favoriteGroup, string level, double discount, int totalSpent = 0)
                    : Customer(name, age, gender, customerId, handphoneNumber, favoriteGroup), Membership(level, discount) {
                        this->totalSpent = totalSpent;
        }

        // Setters and Getters
        void setTotalSpent(int total) {
            this->totalSpent = total;
        }

        int getTotalSpent() {
            return this->totalSpent;
        }

        // For printing VIP customer details
        void printVIPCustomer() {
            cout << left 
                << setw(8)  << getcustomerId()
                << setw(16) << gethandphoneNumber()
                << setw(10) << getName()
                << setw(8)  << getAge()
                << setw(15) << getfavoriteGroup()
                << setw(15) << getLevel()
                << setw(12) << getDiscount()
                << setw(12) << getTotalSpent() << endl;
        }
};

File: PY/Program/Vip.py
from Customer import Customer
from Membership import Membership

class VIPCustomer(Customer, Membership): # Inherit from Customer and Membership
    # Constructor
    def __init__(self, name="", age=0, gender="", customer_id="", handphone_number="", favorite_group="", level="", discount=0.0, total_spent=0):
        Customer.__init__(self, name, age, gender, customer_id, handphone_number, favorite_group)
        Membership.__init__(self, level, discount)
        self._total_spent = total_spent

    # Getter and Setter methods for total_spent
    def set_total_spent(self, total):
        self._total_spent = total

    def get_total_spent(self):
        return self._total_spent    

    # Method to print VIP customer details
    def print_vip_customer(self):
        print(f"{self.get_customer_id():<8}"
            f"{self.get_handphone_number():<16}"
            f"{self.get_name():<10}"
            f"{str(self.get_age()):<8}"
            f"{self.get_favorite_group():<15}"
            f"{self.get_level():<15}"
            f"{(self.get_discount()):<12}"
            f"{self.get_total_spent():<10}")
